packages/udaru-core/lib/hooks.js

```diff
@@ -1,84 +1,101 @@
-      // Call the handler
-      const handlerValue = handler(error, args, result, next)
-
-      // Handle promise returns
-      if (handlerValue && typeof handlerValue.then === 'function') handlerValue.then(next).catch(next)
-    },
-    done // Finish the call
-  )
+module.exports = function buildHooks () {
+  const registered = {}
 
-  return promise
-}
+  function toArray (input) {
+    const output = new Array(input.length)
 
-function wrapWithCallback (name, original, args) {
-  const originalCallback = args.pop() // Save the original callback
-  const originalArgs = Array.from(args) // Save the original arguments
+    for (let i = 0; i < output.length; i++) {
+      output[i] = input[i]
+    }
 
-  args.push((error, ...result) => { // Add the new callback to the udaru method
-    runHook(name, error, originalArgs, result, (err) => { // Run all hooks
-      originalCallback(error || err, ...result) // Call the original callback with the error either from the udaru method or from one of the hooks
-    })
-  })
+    return output
+  }
 
-  // Call the original method
-  original(...args)
-}
+  function runHandlers (name, error, args, results, done) {
+    const hooks = registered[name]
+    const hooksNum = hooks.length
+    let finished = 0
 
-function wrapWithPromise (name, original, args) {
-  let hooksExecuted = false
+    for (let i = 0; i < hooksNum; i++) {
+      hooks[i](error, args, results, err => {
+        finished++
 
-  return original(...args) // Execute the original method
-    .then((result) => {
-      // Now execute hooks
-      hooksExecuted = true
-      return runHook(name, null, args, result).then(() => result)
-    })
-    .catch(error => {
-      // Hooks are already executed, it means they threw an error, otherwise it comes from the original method
-      const promise = hooksExecuted ? Promise.resolve() : runHook(name, error, args, null)
+        if (!done || (!err && finished < hooksNum)) return // The callback has been called or other calls are pending, return
+
+        done(err)
+        done = null
+      })
+    }
+  }
 
-      // Once hooks execution is completed, return any error
-      return promise.then(() => Promise.reject(error))
+  function wrapWithCallback (name, original, args) {
+    const originalCallback = args.pop() // Save the original callback
+
+    // Call the original method with a new callback which will invoke runHook
+    original.apply(this, args.concat(function () {
+      const cbArgs = Array.prototype.slice.call(arguments)
+
+      // The setImmediate call here otherwise any error in the originalCallback will trigger an unhandledRejection
+      runHandlers(name, cbArgs[0], args, cbArgs.slice(1), err => {
+        if (err) cbArgs[0] = err
+
+        originalCallback.apply(null, cbArgs)
+      })
+    }))
+  }
+
+  function wrapWithPromise (name, original, args) {
+    return new Promise((resolve, reject) => {
+      return original.apply(this, args) // Execute the original method
```

Comment:
Can you please clarify this? 
I'm returning the function because I don't want break the chain of the caller. But I cannot simply do `return original.apply(...)` as the runHandlers uses callbacks.

Reply:
You are controlling the outer promise via `resolve, reject`, not with the chain.
Another option is to work only with the chain of promises, like so:

```js
return original.apply(this, args)
  .then(res => {
    return new Promise(...)
  })
  .catch(err => {
    return new Promise(..)
  })
```